Framework/PythonInterface/plugins/algorithms/FindGoniometerFromUB.py

```diff
@@ -0,0 +1,293 @@
+        dPhiRef = phiRef - phiRef[0]
+        dPhiRef = dPhiRef + np.ceil(-dPhiRef / 360) * 360
+
+        # calculate the rotation matrix that maps the UB of the first run onto subsequent UBs
+        for irun in range(1, len(ubFiles)):
+            chi, phi, u = self.getGonioAngles(matUB[irun], zeroUB, omega[irun])
+            # phi relative to first run in RH/LH convention of user
+            # check if phi and chi are not within tolerance of expected
+            if abs(chi - chiRef) > chiTol and abs(phi - dPhiRef[irun]) > phiTol:
+                # generate predicted UB to find axes permutation
+                self.log().information("The following UB\n{}\nis not consistent with the reference, attempting to "
+                                       "find an axes swap/inversion that make it consistent.")
+                # nominal goniometer axis
+                gonio = self.getR(omegaHand * dOmega, [0, 0, 1]) @ self.getR(-chiRef, [1, 0, 0]) @ [0, 0, 1]
+                predictedUB = self.getR(omega[irun], [0, 0, 1]) @ self.getR(dPhiRef[irun], gonio) @ zeroUB
+                # try a permutation of the UB axes (as in TransformHKL)
+                # UB' = UB M^-1
+                # HKL' = M HKL
+                minv = np.linalg.inv(matUB[irun]) @ predictedUB
+                minv = self.getSignMaxAbsValInCol(minv)
+                # redo angle calculation on permuted UB
+                matUB[irun] = matUB[irun] @ minv
+                chi, phi, u = self.getGonioAngles(matUB[irun], zeroUB, omega[irun])
+
+            if abs(chi - chiRef) <= chiTol and abs(phi - dPhiRef[irun]) <= phiTol:
+                # save the consistent UB to the ubPath
+                _, nameUB = path.split(ubFiles[irun])
+                newUBPath = path.join(saveDir, nameUB[:-4] + '_consistent' + nameUB[-4:])
+                self.writeConsistentUB(ubFiles[irun], newUBPath, matUB[irun])
+                # populate row of table
+                phi2print = phiHand * (phi + phiRef[0])
+                phi2print = phi2print + np.ceil(-phi2print / 360) * 360
+                nextRow = {'Run': nameUB[:-4],
+                           'Chi': chi,
+                           'Phi': phi2print,
+                           'GonioAxis': V3D(u[0], u[1], u[2])}
+                gonioTable.addRow(nextRow)
+            else:
+                warnings.warn("WARNING: The UB {0} cannot be made consistent with the reference UB. "
+                              "Check the goniometer angles and handedness supplied "
+                              "and the accuracy of reference UB.".format(ubFiles[irun]))
+
+        # assign output
+        self.setProperty("OutputTable", gonioTable)
+        # complete progress
+        prog_reporter.report(3, "Done")
+
+    def readUB(self, fpath):
+        """
+        :param fpath: path to UB files exported with SaveIsawUB
+        :return: 3x3 UB matrix
+        """
+        ub = np.loadtxt(fpath, skiprows=0, max_rows=3).T
```

Comment:
It would appear that the `max_rows` argument was new in numpy version 1.16 and on RHEL/Ubuntu we only have 1.12/1.13 respectively. :(
https://stackoverflow.com/a/32560011 has a workaround using `itertools`.